shaders/src/main/glsl/samples/310es/colorgrid_modulo_intfunctions.frag

```diff
@@ -0,0 +1,60 @@
+#version 310 es
+
+/*
+ * Copyright 2019 The GraphicsFuzz Project Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+precision highp float;
+
+layout(location = 0) out vec4 _GLF_color;
+
+uniform vec2 resolution;
+
```

Comment:
It looks like result of most the built-ins here will be evaluated at compile-time. In particular, `bitfieldInsert(800, 0, 0, 0)` is a cool function, but that will probably just be trivially replaced with `800` at compile-time.

We could try passing a loop counter (`i` or `j`) to a built in. I came up with this:

```glsl
    // within the loop body:
    int another_i = i;
    // Set the 16th bit.
    another_i = bitfieldInsert(another_i, 1, 16, 1);
    int sixteenthBit = int(0x010000 * int(injectionSwitch.y));
    // Unset the 16th bit
    another_i = another_i & ~sixteenthBit;
    // now use another_i instead of i within the loop...but still increment i!
```

This can't be simplified completely at compile-time. We could just modify `i` directly, since we don't actually change its value in the end, but it might be best not to do that so that the compiler can still easily see that the loop iterates 800 times, and so it might be able to do some extra optimizations (although 800 iterations is probably too many for the compiler to bother trying to unroll the loop). 

Of course, doing something like this *everywhere* might complicate things too much but, in general, passing literals to built-ins will probably just get optimized away (but it is still good to do it sometimes).

Reply:
Just for fun, I changed it to `bitfieldInsert(800, i, 0, 0)`; I'm not sure if the compiler can optimize that at compile time when it's checked every loop and the loop isn't unrolled.